Create data layer for Marketplace reviews

gcsecsey opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/4556

## Proposed Changes

* add methods to use the new review API endpoints
* add some elements to the marketplace test site to be able to test requests

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply the patch in D129780-code and sandbox `public-api`
* Go to `/marketplace/test/:siteID`
* Use the new card on the top of the page to add/update/delete reviews
* Check that the query on the top of the page is refetched after each operation

![CleanShot 2023-11-22 at 13 05 08@2x](https://github.com/Automattic/wp-calypso/assets/11555574/1e77bf66-a1f9-49e4-a02a-a701499237c1)

 
## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

gcsecsey committed:
add query for all reviews

gcsecsey committed:
add mutation wrappers for create/update/delete methods

gcsecsey committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into reviews-data-layer

flootr commented on client/data/marketplace/use-marketplace-reviews.ts:
```diff
@@ -0,0 +1,134 @@
+	return useMutation( {
+		mutationFn: createReview,
+		onSuccess: () => {
+			queryClient.invalidateQueries( queryKeyBase );
```
Please use object syntax:
```ts
invalidateQueries( { queryKey: queryKeyBase } )
```
The upcoming v5 of RQ removes any other signatures for `useQuery` and friends except for object syntax. The same is true for all other usages in this PR. (see the [migration guide](https://tanstack.com/query/v5/docs/react/guides/migrating-to-v5#supports-a-single-signature-one-object) on which methods are affected)

gcsecsey replied:
Updated, thanks. I wondered if there would be an easy way to move the invalidateQueries to a separate hook, then reuse that hook in the other ones:

```ts
const useInvalidateQueries = ( options ) => {
    const queryClient = useQueryClient();
	return useMutation( {
		onSuccess: () => {
			queryClient.invalidateQueries( { queryKey: queryKeyBase } );
		},
                ...options
	} );
}

export const useCreateMarketplaceReviewMutation = () => {
	return useInvalidateQueries( {
		mutationFn: createReview
	} );
};
```

I tried this out, but was unable to type the query invalidator hook properly. Would RQ still be able to infer the correct return types with a similar indirection somehow?

flootr replied:
I don't think that's an abstraction worth doing at this point tbh. Also I think `useInvalidateQueries` is not a good name as it's very specific and you might end up wanting to reuse more than just that. v5 introduces [`queryOptions()`](https://tanstack.com/query/v5/docs/react/typescript#typing-query-options) function which should make it easier to extract query options – although I'm not sure if it'd help in this particular case.

flootr commented on client/data/marketplace/use-marketplace-reviews.ts:
```diff
@@ -0,0 +1,134 @@
+export const useMarketplaceReviews = (
+	{ productType, pluginSlug }: ProductProps,
+	{ enabled = true, staleTime = BASE_STALE_TIME, refetchOnMount = true }: UseQueryOptions = {}
+): UseQueryResult< any > => {
```
Instead of using `any` it's recommended to type your fetch function `fetchMarketplaceReviews` a few lines above and let RQ infer from that. That way you don't have to specify a `ReturnType` here.

gcsecsey replied:
Great point, thanks, I added the return types.

flootr commented on client/data/marketplace/use-marketplace-reviews.ts:
```diff
@@ -0,0 +1,134 @@
+	reviewId: number;
+} & ProductProps;
+
+const fetchMarketplaceReviews = ( productType: ProductType, productSlug: string ) => {
```
If you add a `ReturnType` to this function RQ will be able to infer the type within the query hook. See my comment on L94. Something like

```ts
type MarketplaceReviewsResponse = {
  /* ... */
}

const fetchMarketplaceReviews = ( productType: ProductType, productSlug: string ): Promise< MarketplaceReviewsResponse > => {
  /* ... */
}

flootr commented on client/data/marketplace/use-marketplace-reviews.ts:
```diff
@@ -0,0 +1,134 @@
+
+export const useMarketplaceReviews = (
+	{ productType, pluginSlug }: ProductProps,
+	{ enabled = true, staleTime = BASE_STALE_TIME, refetchOnMount = true }: UseQueryOptions = {}
```
If you don't want to expose the whole API surface for `useQuery` I'd generally advise to narrow it down to just the properties you know you're going to use – you already did that, however, `UseQueryOptions` used like that is not the correct type to use. Some alternatives:

```ts
type MarketplaceReviewsQueryOptions = Pick<
  UseQueryOptions,
  'enabled' | 'staleTime' | 'refetchOnMount'
>;
```

If you type your fetch function like I suggested:

```ts
type MarketplaceReviewsQueryOptions = Pick<
  UseQueryOptions< MarketplaceReviewsResponse >,
  'enabled' | 'staleTime' | 'refetchOnMount'
>;
```

As soon as you expose any properties which make use of `data` using the generic is recommended because otherwise it won't receive the correct `data` type (`refetchOnMount` accepts a function which is given the query which also contains `data`).

In any case what you should do because it will type error in v5 is omitting `queryKey` from your options type when using `UseQueryOptions` like that (not needed when you do one of the above):

```ts
const options: Omit< UseQueryOptions, 'queryKey' > = { /* ... */ }
// or
const options: Omit< UseQueryOptions< MarketplaceReviewsResponse >, 'queryKey' > = { /* .. */ }
```

flootr reviewed (commented):
👋🏻 drive by. You haven't asked for a review yet but I wanted to chime in regardless and make some suggestions regarding our usage of RQ in Calypso. In addition to my comments below we have a soft convention to name your hooks like `useXMutation` and `useYQuery` to make it easier to differentiate with non-RQ hooks.

gcsecsey committed:
update types

gcsecsey committed:
use DELETE method for deletion

gcsecsey committed:
update hook names

gcsecsey committed:
update hook names

gcsecsey committed:
Merge branch 'reviews-data-layer' of github.com:Automattic/wp-calypso into reviews-data-layer

gcsecsey commented:
Thanks a lot @flootr, for the detailed review and guidance with RQ! 🎉 I updated the files based on your suggestions, please give it another look!

gcsecsey committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into reviews-data-layer

gcsecsey commented:
I've updated the test steps with the new diff from @gavande1, enabling the DELETE method on this PR.

I've also fixed the CI build, so this is now ready for a review from @Automattic/bespin

epeicher reviewed (approved):
I have tested it, and it works as advertised. I can add 
![CleanShot 2023-11-23 at 13 10 56@2x](https://github.com/Automattic/wp-calypso/assets/3519124/eea48f72-e990-473b-b342-a10f10b768f1)
, update 
![CleanShot 2023-11-23 at 13 11 10@2x](https://github.com/Automattic/wp-calypso/assets/3519124/dd789c90-28b8-41cc-94fc-319c11079f46)
 and delete 
![CleanShot 2023-11-23 at 13 11 33@2x](https://github.com/Automattic/wp-calypso/assets/3519124/2d39aad1-5da7-4f62-8572-741735beb3c6)
comments

I have a question that can be handled as a follow-up. Should we return any errors from the hooks so the UI can display some error notification?

gcsecsey commented:
> I have a question that can be handled as a follow-up. Should we return any errors from the hooks so the UI can display some error notification?

The errors are returned from the hooks currently, eg. with an existing review, trying to create a new review from the `/marketplace/test` page will return the API error in the `error` property of the useMutation hook:
![CleanShot 2023-11-23 at 12 24 17@2x](https://github.com/Automattic/wp-calypso/assets/11555574/95f56ac2-f588-49b2-abad-692481994c2d)

We can use this response in the frontend tickets later to show the appropriate error messages on the UI.

epeicher commented:
> The errors are returned from the hooks currently, eg. with an existing review, trying to create a new review from the `/marketplace/test` page will return the API error in the `error` property of the useMutation hook:

Oh, I see. I was focusing in the `onError` callback in the `useMutation` [here](https://github.com/Automattic/wp-calypso/pull/84379/files#diff-619bf331019437d595f24ba545fe9419413a24aba100aa59baffbc37d4f93ae0R158) but I see they are returned and can be used as you suggest. That is a great option. Thanks for the answer!

gcsecsey committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into reviews-data-layer
